### ISSUE TITLE
Critical: better salt for password derivation

### ISSUE THREAD

**Issue by JosePisco** [Cosmian/kms#124]
Passwords can be derived into secure keys via the `derive_key_from_password()` method in `crate/utils`.
There is an issue where one would not find back the same key from a password twice and because the salt is not saved.
This could be fixed by adding an `Option(&[u8])` parameter to `derive_key_from_password()` to both FIPS (PBKDF2 with HMAC256) and non-FIPS (Argon2).

Adding support for non-deterministic salt to Argon2 is primordial since for the moment the salt is a constant string, which means that two identical password derive to the same key and this need to be avoided.

Don't forget to ad tests to `derive_key_from_password()` with optional salt.

**Comment by JosePisco** [comment]
Test `password_wrap_import_test()` has been fips altered for `grnx-fix-fips` branch to be merged without overextending changes. Remember to remove `#[cfg(not(feature = "fips"))]` flag.

**Comment by JosePisco** [comment]
After commit a8b8a068f5d0fbf5091586627afd5b618dc59025, I took the decision to take the salt as a parameter defined as high as possible in the recursion calls. An example is the `password_wrap_import_test()` which gives an hexstr for the salt to use.

I'm not really a fan of this since it means (does it ?) the user has to handle salt + we are not sure the salt he will provide will be "FIPS-compliant-generated" (i.e. 128 bits of cryptographically secure randomness).

I have added the salt as a return value of `derive_key_from_password()` in case no salt is provided in input so that one can use it later for identical derivation for example if you want to derive the same key from a password twice.

There are two options here:
- first is we allow users to specify a salt and admit they are doing it well (which I don't feel is a great idea)
- we forbid them from specifying one and we handle everything ourselves. This means that we will have to internally store salts so we know when we need to generate a new one / use one we stored previously.
